### DISCUSSION TITLE
idea: `useContractEvent` should accept topics

### DISCUSSION THREAD

**Discussion by kdembler** [wevm/wagmi#1378] [Ideas]
### Current Behavior

When using `useContractEvent`, since v0.7, if you provide ABI with const type assertion, `eventName` will only accept a string that matches one of contract's events names. Prior to v0.6 it was also possible to provide `TypedEventFilter` for that field, basically allowing anything that ethers `contract.on` accepts.

### Expected Behavior

I would expect the library to also accept `TypedEventFilter` created with `contract.filters.Event()` so that I can fine-tune event subscription. Currently, I'm forced to listen to all the events, with no ability to filter based on specific topics. I think `eventName: contract.filters.Event(/* ... */)` should be accepted and provide valid `listener` typings.

**Comment by jxom** [comment]
Absolutely agree we should support this! It is on the roadmap (we would like to try find a friendlier API for filters than just an array input, hence why we didn't include the proper types for it post v0.7) – planning to work on it soon!

For now, might have to explicitly use Ethers to achieve this.

Just going to move this one to discussions under "Ideas" as this is more of a feature request than an issue.

**Comment by vseehausen** [comment]
Hey, what is the status? It is recommended to use viem for this?

**Reply by kdembler** [reply]
No support in `wagmi` yet but it's really easy to do with `viem` after 1.0:
```ts
import { useAccount, usePublicClient } from 'wagmi';

function useSubscribeEvent() {
  const publicClient = usePublicClient();
  const { address } = useAccount();

  useEffect(() => {
    if (!publicClient || !address) {
      return;
    }
    const unsub = publicClient.watchContractEvent({
      address: CONTRACT_ADDRESS,
      abi: CONTRACT_ABI,
      eventName: 'MyEvent',
      args: {
        trader: address,
      },
      onLogs: async (logs) => {
        for (const log of logs) {
          console.log(log);
        }
      },
    });
    return () => unsub();
  }, [publicClient, address]);
}
```
And it's all typesafe!

**Reply by zkSoju** [reply]
> No support in `wagmi` yet but it's really easy to do with `viem` after 1.0:
> 
> ```ts
> import { useAccount, usePublicClient } from 'wagmi';
> 
> function useSubscribeEvent() {
>   const publicClient = usePublicClient();
>   const { address } = useAccount();
> 
>   useEffect(() => {
>     if (!publicClient || !address) {
>       return;
>     }
>     const unsub = publicClient.watchContractEvent({
>       address: CONTRACT_ADDRESS,
>       abi: CONTRACT_ABI,
>       eventName: 'MyEvent',
>       args: {
>         trader: address,
>       },
>       onLogs: async (logs) => {
>         for (const log of logs) {
>           console.log(log);
>         }
>       },
>     });
>     return () => unsub();
>   }, [publicClient, address]);
> }
> ```
> 
> And it's all typesafe!

This would be very useful for useContractEvent hooks!
